### PR TITLE
add fieldName

### LINKED ISSUE #162: Multiple cookies
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Add support for different "instances" of the session plugin, with different cookie names and cookie options (different path/expiration date). This could be accomplished by allowing to specify the decorated field name (`request.session` by default) in plugin options.

Please keep in mind this will require changing the way types work, as the `SessionData` is currently set for the entire `@fastify/secure-session` module

### Motivation

I need to keep a separate session for short-term (data expiring after closing tab) and long-term storage (user login token)

### Example

```ts
const SecureSession = require('@fastify/secure-session');

fastify.register(SecureSession, {
  cookieName: 'long-term-cookie',
  cookie: {
    path: '/'
    expires: new Date('2137-01-01'),
  },
  field: 'longTermSession'
});
fastify.register(SecureSession, {
  cookieName: 'short-term-cookie',
  cookie: {
    path: '/'
    expires: undefined,
  },
  field: 'shortTermSession'
});

fastify.get('/', (request) => {
  console.log(request.longTermSession);
  console.log(request.shortTermSession);
});
```

### PR DESCRIPTION
**Pull request by Uzlopak**
Closes #162 

Any better name than fieldName?

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by dominik-korsa** [comment]
There need to be a way for users to specify different SessionData for each instance, I'll try to implement this using generics

**Comment by Uzlopak** [comment]
You are right, it already has decorated request with functions...

**Comment by dominik-korsa** [comment]
I think the default `FastifyRequest` interface declaration merge and the `SessionData` interface should be removed. That would be a breaking change (although it would only affect TypeScript compilation and not runtime, so maybe you will find it acceptable). `Session` could now take a template argument for `SessionData`, and TypeScript users will be expected to add the `FastifyRequest` declaration themselves, like this:
```ts
interface FooSessionData {
  foo: string;
}
interface BarSessionData {
  bar: number;
}

declare module "fastify" {
  interface FastifyRequest {
    foo: Session<FooSessionData>;
    bar: Session<BarSessionData>;
  }
}
```

**Comment by Uzlopak** [comment]
I think you are missing the point. the typings are the least of the issues.

**Comment by mcollina** [comment]
We use `name` in https://github.com/fastify/fastify-mongodb.

File: README.md [hunk 1]
```diff
@@ -34,6 +34,8 @@ const path = require('path')
 fastify.register(require('@fastify/secure-session'), {
   // the name of the session cookie, defaults to 'session'
   cookieName: 'my-session-cookie',
+  // the name of the attribute decorated on the request-object, defaults to 'session'
+  fieldName: 'my-session',
```
Comment (Eomm) [inline README.md:38]: We don't have a "standard" for it.

`name` is too generic IMHO

`fastify-env` uses `confKey`, maybe `sessionKey` could fit better?
https://github.com/fastify/fastify-env#usage

**Comment by Uzlopak** [comment]
The issue is more, that we decorate fastify and Request with decodeSecureSession. So we can not attach the plugin multiple times.

**Comment by Uzlopak** [comment]
closing this, as it is not possible to have two decodeSecureSession decorators

**Pull request closed without merging**